library/bignum_mod_raw.c

```diff
@@ -125,16 +123,45 @@ void mbedtls_mpi_mod_raw_fix_quasi_reduction(mbedtls_mpi_uint *X,
     (void) mbedtls_mpi_core_add_if(X, N->p, N->limbs, (unsigned) c);
 }
 
-#endif /* MBEDTLS_TEST_HOOKS */
 
 void mbedtls_mpi_mod_raw_mul(mbedtls_mpi_uint *X,
                              const mbedtls_mpi_uint *A,
                              const mbedtls_mpi_uint *B,
                              const mbedtls_mpi_mod_modulus *N,
                              mbedtls_mpi_uint *T)
 {
-    mbedtls_mpi_core_montmul(X, A, B, N->limbs, N->p, N->limbs,
-                             N->rep.mont.mm, T);
+    /* Standard (A * B) multiplication stored into pre-allocated T
+     *  buffer of fixed limb size of (2N + 1).
+
+     *  The space may not not fully filled by when
+     *  MBEDTLS_MPI_MOD_REP_OPT_RED is used, with most
+     *  curves using (2N) limbs.
+     *
+     *  The 521-bit Weierstrass curve is the only
+     *  that which requires a limb size of (2N + 1). */
+    const size_t T_limbs = (N->bits == 521) ?
+                           BITS_TO_LIMBS(N->bits * 2) + 1 :
```

Comment:
This is used in the mbedtls_ecp_mod_p521_raw method ([link](https://github.com/Mbed-TLS/mbedtls/blob/development/library/ecp_curves.c#L5352-L5354)). 

I am a bit sceptical on the need for the zero input check in that curve and which corner case it may be required. 

I do appreciate that it is meant for catching zero padding inputs, but it would allow the input check to pass if `X_limbs < P521_WIDTH` and the random memory allocated at the next addresses meets that requirement `X[2 * P521_WIDTH - 1] != 0`

Reply:
I don't think that `mbedtls_ecp_mod_p521_raw()` needs that extra limb, we should just adjust the check in `mbedtls_ecp_mod_p521_raw()` and the `expected_width` in `ecp_mod_p521()` accordingly.